lib/experimental/colortext.nim

```diff
@@ -0,0 +1,765 @@
+  while 0 <= idx:
+    yield text.runes[idx]
+    dec idx
+
+func `==`*(s1, s2: ColStyle): bool =
+  let check =
+    if s1.use8Bit:
+      s1.fg8 == s2.fg8 and s1.bg8 == s2.bg8
+
+    else:
+      s1.fg == s2.fg and s1.bg == s2.bg
+
+
+  s1.style == s2.style and
+  s1.use8Bit == s2.use8Bit and check
+
+func contains*(ps: ColStyle, s: Style): bool =
+  ps.style.contains(s)
+
+func initColStyle*(
+    fg: ForegroundColor = fgDefault,
+    bg: BackgroundColor = bgDefault,
+    style: set[Style] = {}
+  ): ColStyle =
+  ColStyle(use8Bit: false, fg: fg, bg: bg, style: style)
+
+const defaultColStyle* = initColStyle()
+
+func default*(rune: typedesc[ColRune]): ColRune =
+  ColRune(style: defaultColStyle)
+
+func default*(style: typedesc[ColStyle]): ColStyle =
+  ColStyle(use8Bit: false, fg: fgDefault, bg: bgDefault)
+
+func initStyleBg*(term: TermColor8Bit): ColStyle {.inline.} =
+  ColStyle(use8Bit: true, bg8: term)
+
+func initStyleFg*(term: TermColor8Bit): ColStyle {.inline.} =
+  ColStyle(use8Bit: true, fg8: term)
+
+func initStyle*(fg, bg: TermColor8Bit): ColStyle {.inline.} =
+  ColStyle(use8Bit: true, fg8: fg, bg8: bg)
+
+func uc*(s: static[string]): Rune = runeAt(s, 0)
+func `+`*(bg: TermColor8BitBg, fg: TermColor8BitFg): ColStyle =
+  ColStyle(use8Bit: true,
+               fg8: TermColor8Bit(fg.int),
+               bg8: TermColor8Bit(bg.int))
+
+func toColRune*(rune: Rune, style: ColStyle): ColRune =
+  ColRune(rune: rune, style: style)
+
+func `+`*(style: set[Style], bg: BackgroundColor): ColStyle =
+  ColStyle(use8Bit: false, bg: bg, style: style)
+
+func `+`*(style: set[Style], fg: ForegroundColor): ColStyle =
+  ColStyle(use8Bit: false, fg: fg, style: style)
+
+func `+`*(fg: TermColor8BitFg, bg: TermColor8BitBg): ColStyle =
+  ColStyle(use8Bit: true,
+               fg8: TermColor8Bit(fg.int),
+               bg8: TermColor8Bit(bg.int))
+
+func `+`*(bg: BackgroundColor, fg: ForegroundColor): ColStyle =
+  ColStyle(use8bit: false, fg: fg, bg: bg)
+
+func `+`*(fg: ForegroundColor, bg: BackgroundColor): ColStyle =
+  ColStyle(use8bit: false, fg: fg, bg: bg)
+
+func `+`*(style: sink ColStyle, bg: BackgroundColor): ColStyle =
+  result = style
+  result.bg = bg
+
+func `+`*(style: sink ColStyle, fg: ForegroundColor): ColStyle =
+  result = style
+  result.fg = fg
+
+func `+`*(style: sink ColStyle, s: Style): ColStyle =
+  result = style
+  result.style.incl s
+
+func `+`*(ch: char, style: ColStyle): ColRune =
+  toColRune(Rune(ch), style)
+
+func `+`*(rune: Rune, style: ColStyle): ColRune =
+  toColRune(rune, style)
+
+func `+`*(str: string, style: ColStyle): ColText =
+  for ch in runes(str):
+    result.runes.add toColRune(ch, style)
+
+func `+`*(str: string, fg: ForegroundColor): ColText =
+  str + (fg + bgDefault)
+
+func `+`*(rune: Colrune, style: ColStyle): ColRune =
+  result = rune
+  result.style = style
+
+func `+`*(text: ColRuneLine, style: ColStyle): ColRuneLine =
+  for ch in text:
+    result.add ch + style
+
```

Comment:
so many `+` operators. Nim uses `&` for string concatenation. Why differ here?

Reply:
Because it is not a string concatenation, it is a `text + style` operation. I don't specifically care what operator to use here, bit `&` wasn't used specifically because it is not a concatenation.

Reply:
Please don't use the mathematical `+` operator for this. Using operators this way, was the primary reason many programming languages abandoned operator overloading entirely.

Reply:
What operator do you want to use then? Except for `&` which is used to denote concatenation, which is not what is going on here.

Reply:
`+` is already used for sets, and for `style + style` operations it is exactly what happens - I merge to sets of formatting options.

Using `&` for `text + style` will prevent code like `"red" + fgRed & "blue" + fgBlue` and require `("red" + fgRed) & ("blue" + fgBlue)` for no apparent benefit

`%` might be used for `text % style` it has higher precedence, and already used for string interpolation in `std/strutils`

Reply:
I am not sure yet. I don't know about any established style operators yet. Generally I would prefer a real verb instead of an operator. From my experience in language design it is as important to read something as it is to write it.